### PR TITLE
Add tag sync support for Sonarr and Radarr

### LINKED ISSUE #500: [FEAT]: Sync indexer tags to indexers in apps || Add Application Tags to Indexers
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Is your feature request related to a problem? Please describe

I currently use indexer tags to filter the indexers used for various series. While I can add these in Prowlerr, they don't seem to sync to the apps.

### Describe the solution you'd like

Tags applied in Prowlerr are automatically synced to apps.

### Describe alternatives you've considered

Adding tags manually removes them on next sync.

### Anything else?

Seems odd for this field to be mandatory...

[AB#1579](https://dev.azure.com/Servarr/7ab38f4e-5a57-4d70-84f4-94dd9bc5d6df/_workitems/edit/1579)

**EDIT:** Corrected info on alternatives attempted. Forgot it was in "Add and remove only" sync mode so hadn't updated.

### PR DESCRIPTION
**Pull request by 6cUbi57z**
#### Database Migration
NO

#### Description

Adds tag sync support to Sonarr and Radarr.

- **Off (Default):** Leaves tags in application indexers as they are. Due to the way in which the sonarr and radarr API seem to work, this requires querying the API to get the tags in use, then include those in the update.
- **On:** Overwrites all tags with those configured on the indexer in Prowlarr. Any tags which do not exist will be automatically created.

#### Screenshot (if UI related)

image
image

#### Todos
- [ ] Tests
- [ ] Translation Keys (./src/NzbDrone.Core/Localization/Core/en.json)
- [ ] [Wiki Updates](https://wiki.servarr.com)

#### Issues Fixed or Closed by this PR

* Partially resolves #500

### CODE REVIEW

**Comment by bakerboy448** [comment]
Are using the indexers' tags the best option for this? That can mean a lot of unecessary tags going to an app depending on how one users tags.

App Tags should probably be a different option maybe, no?

**Comment by 6cUbi57z** [comment]
Do you mean to have a separate setting on each indexer which are the tags to transfer? I like the idea to keep them split, but I can see terminology getting very confusing. I already find the tagging to link indexer to proxy and app a bit confusing. I'm not sure it would help matters if indexer tags in prowlarr don't result as indexer tags in apps.

As for unnecessary tags, I'm not sure how many tags people usually have. It should be minimised somewhat as sync is off by default and will only create tags which it needs to sync. Personally I wouldn't care about some unused tags. They're not very visible in the apps and don't affect functionality. That's just my opinion though.

**Comment by 6cUbi57z** [comment]
Changed to WIP I believe there is a bug noticed in thread at https://github.com/Prowlarr/Prowlarr/issues/500#issuecomment-1383262604. When tag sync is off but fully indexer sync is on, the tags will be deleted from the indexer in the app. They should be left as is.

**Comment by 6cUbi57z** [comment]
Updated to preserve existing tags when tag sync is off. PR description updated.

File: src/NzbDrone.Core/Applications/Radarr/Radarr.cs [hunk 1]
```diff
@@ -203,5 +213,41 @@ private RadarrIndexer BuildRadarrIndexer(IndexerDefinition indexer, DownloadProt
 
             return radarrIndexer;
         }
+
+        private HashSet<int> GetAndCreateSonarrTagIdsForIndexer(IndexerDefinition indexer)
```
[the remark this answers is not in the document]
Comment (6cUbi57z) [reply src/NzbDrone.Core/Applications/Radarr/Radarr.cs:217]: Changed to `GetAndCreateApplicationTagIdsForIndexer` in both Sonarr and Radarr to try avoid future copy paste errors and still differentiate between Sonarr/Radarr tag IDs  and Prowlarr Tag IDs.

File: src/NzbDrone.Core/Applications/Sonarr/SonarrV3Proxy.cs [hunk 2]
```diff
@@ -140,6 +142,21 @@ public ValidationFailure TestConnection(SonarrIndexer indexer, SonarrSettings se
...
+            var request = BuildRequest(settings, $"/api/v3/tag/detail", HttpMethod.Get);
+            return Execute<List<SonarrTag>>(request);
+        }
+
+        public SonarrTag CreateTag(SonarrSettings settings, string label)
+        {
+            var request = BuildRequest(settings, $"/api/v3/tag", HttpMethod.Post);
+            request.SetContent(new SonarrTag { Label = label }.ToJson());
+            var tag = Execute<SonarrTag>(request);
+            _logger.Info("Tag '{0}' created or already existed with ID '{2}'.", tag.Label, tag.Id);
```
[the remark this answers is not in the document]
Comment (6cUbi57z) [reply src/NzbDrone.Core/Applications/Sonarr/SonarrV3Proxy.cs]: Fixed here and in Radarr equivalent.

File: src/NzbDrone.Core/Applications/Radarr/RadarrIndexer.cs [hunk 3]
```diff
@@ -48,14 +48,16 @@ public bool Equals(RadarrIndexer other)
             var otherSeedRatio = other.Fields.FirstOrDefault(x => x.Name == "seedCriteria.seedRatio")?.Value == null ? null : (double?)Convert.ToDouble(other.Fields.FirstOrDefault(x => x.Name == "seedCriteria.seedRatio").Value);
             var seedRatioCompare = seedRatio == otherSeedRatio;
 
+            var tagsEqual = (Tags ?? new HashSet<int>()).Equals(other.Tags ?? new HashSet<int>());
```
[the remark this answers is not in the document]
Comment (6cUbi57z) [reply src/NzbDrone.Core/Applications/Radarr/RadarrIndexer.cs]: I think general practice in C# is for "compare" (e.g. `IComparer`) to be an int indicating if it is greater, equal or less than. "Equal" is generally a boolean. I think this also reads better as a bool true clearly indicates it is true that it is equal...

...however, given the other variables, I'm fine with changing this for consistency so have updated it.

**Comment by 6cUbi57z** [comment]
Good catch, but I believe it is not quite that simple now. If I've worked this out correctly, that line will mean that a tag being removed in prowlarr won't be removed in the other application as that line will re-add it after the new indexer is built.

Potential options I can see are:

- Leave the retain tags line as is so that Prowlarr tags are only ever added and never removed.
- Remove the retain tags line to have it controlled by BuildSonarrIndexer. This would provide full sync, or retain tags if the sync tags setting is off (the code for this can be made more efficient to avoid multiple API calls).
- Update the code to consider tags which Prowlarr is aware of. For tags existing in both apps, perform a full sync (add and remove), for tags within applications which are NOT available in Prowlarr, leave them as they are.
- Update the tag sync setting so that it can be switched between the above options. This may get confusing as there will then be 2 "sync level" settings.

Given the discussions around this PR and the linked issue, the last option may be best to offer the most flexibility.

I started working on the tests but this and the tests might take a while to resolve so I'll put this back to WIP for now.

**Comment by bakerboy448** [comment]
@6cUbi57z still open / getting to eventually or close?

**Comment by 6cUbi57z** [comment]
@bakerboy448 I don't think I'm going to get to it and am not using Prowlarr at the moment so I'll close it. I'll leave my branch around if anyone else wants to pick it up.

**Pull request closed without merging**